### PR TITLE
In-memory Engine: decoupling raftstore and IME

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: Ref https://github.com/tikv/tikv/issues/17181

What's Changed:

```commit-message
* Do not disable snapshot cache when IME is enabled

    Since `async_in_memory_snapshot` always takes an in-memory engine
    snapshot regardless of whether the RocksDB snapshot is cached, we can
    safely revert #16863 and address @tonyxuqqi's concern raised in
    https://github.com/tikv/tikv/pull/16863#discussion_r1571496020

* Remove SnapshotContext from LocalReadRouter

    This commit partially reverts #16181.

* Remove SnapshotContext from KvEngine::snapshot

    This commit partially reverts #16163. By removing `SnapshotContext`
    from `KvEngine::snapshot`, raftstore is no longer coupled with the
    In-Memory Engine.
```

### Check List

Tests 

- Unit test
- Integration test

### Release note

```release-note
None
```

### CODE REVIEW

File: components/hybrid_engine/src/observer/snapshot.rs [hunk 1]
```diff
@@ -36,11 +37,17 @@ impl HybridSnapshotObserver {
 }
 
 impl SnapshotObserver for HybridSnapshotObserver {
-    fn on_snapshot(&self, ctx: SnapshotContext, sequence_number: u64) -> Box<dyn ObservedSnapshot> {
+    fn on_snapshot(
```
Comment (glorv) [inline components/hybrid_engine/src/observer/snapshot.rs:40]: Based on previous discussion, shall we remove this `HybridSnapshotObserver`? Or you decided to do it in another PR?

Comment (overvenus) [reply components/hybrid_engine/src/observer/snapshot.rs:40]: I want to do it in another PR, because it requires more tests.

**Review by SpadeA-Tang — APPROVED** [review]

**Review by glorv — APPROVED** [review]
